<div style="border:solid Chocolate 2px; padding: 40px">

Привет, меня зовут Александр Куимов. Я буду ревьюером твоего проекта. Ты можешь обращаться ко мне на "ты"😏 Надеюсь, тебя также не смутит, если я буду обращаться к тебе на "ты", но если это неудобно, обязательно скажи об этом!

Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 

Ты также можешь реагировать на мои комментарии своими по шаблону, показанному чуть ниже. Это нужно, чтобы не создавалась путаница😉

Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
    
<br/>    
<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b>👋:</b> В такой цветовой ячейке я прошу тебя оставлять свои комментарии. Если исправляешь проект на второй итерации и выше, не забывай пожалуйста указывать номер итерации, например, "Комментарий студента v.2".
</div> 

<br/>    
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самому найти и исправить ее. На реальной работе твой руководитель будет поступать также, и я пытаюсь подготовить тебя именно к работе датасаентистом. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку!🤓

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Лемматизация</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Logistic-Regression" data-toc-modified-id="Logistic-Regression-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Logistic Regression</a></span></li><li><span><a href="#Random-Forest" data-toc-modified-id="Random-Forest-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Random Forest</a></span></li><li><span><a href="#Decision-Tree" data-toc-modified-id="Decision-Tree-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Decision Tree</a></span></li><li><span><a href="#LightGBM" data-toc-modified-id="LightGBM-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>LightGBM</a></span></li></ul></li><li><span><a href="#Анализ-моделей" data-toc-modified-id="Анализ-моделей-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Анализ моделей</a></span></li><li><span><a href="#Тестирование-лучшей-модели" data-toc-modified-id="Тестирование-лучшей-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование лучшей модели</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

 <div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Вижу твое добавленное описание проекта. Молодец! Это поможет тебе расставлять акценты в выводах

In [1]:
import pandas as pd

import re
import nltk
from tqdm import tqdm
from nltk.corpus import stopwords as nltk_stopwords
import spacy

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from lightgbm import LGBMClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.metrics import f1_score

import warnings
warnings.filterwarnings('ignore')

RANDOM_STATE=12345

## Подготовка

In [2]:
try:
    df = pd.read_csv('toxic_comments.csv')
except:
    df = pd.read_csv('/datasets/toxic_comments.csv')

df = df[['text', 'toxic']]
df

text  toxic
0       Explanation\nWhy the edits made under my usern...      0
1       D'aww! He matches this background colour I'm s...      0
2       Hey man, I'm really not trying to edit war. It...      0
3       "\nMore\nI can't make any real suggestions on ...      0
4       You, sir, are my hero. Any chance you remember...      0
...                                                   ...    ...
159287  ":::::And for the second time of asking, when ...      0
159288  You should be ashamed of yourself \n\nThat is ...      0
159289  Spitzer \n\nUmm, theres no actual article for ...      0
159290  And it looks like it was actually you who put ...      0
159291  "\nAnd ... I really don't think you understand...      0

[159292 rows x 2 columns]

### Лемматизация

In [3]:
df['text'] = df['text'].str.lower()

In [4]:
def prepare_text(text):
    text = re.sub(r'(?:\n|\r)', ' ', text)
    text1 = re.sub(r'[^a-zA-Z]', ' ', text)
    
    return " ".join(text1.split())
    
df['text'] = df['text'].apply(prepare_text)

In [5]:
new_corpus = []
nlp = spacy.load('en_core_web_sm')
for doc in tqdm(nlp.pipe(df['text'], batch_size=70, n_process=-1, disable=['parser', 'ner']), total=len(df['text'])):
    word_list = [tok.lemma_ for tok in doc]
    new_corpus.append(' '.join(word_list))

df['lemm_text'] = new_corpus

100%|█████████████████████████████████████████████████████████████████████████| 159292/159292 [14:22<00:00, 184.75it/s]


In [6]:
df = df.dropna(subset=['lemm_text'])
df

text  toxic  \
0       explanation why the edits made under my userna...      0   
1       d aww he matches this background colour i m se...      0   
2       hey man i m really not trying to edit war it s...      0   
3       more i can t make any real suggestions on impr...      0   
4       you sir are my hero any chance you remember wh...      0   
...                                                   ...    ...   
159287  and for the second time of asking when your vi...      0   
159288  you should be ashamed of yourself that is a ho...      0   
159289  spitzer umm theres no actual article for prost...      0   
159290  and it looks like it was actually you who put ...      0   
159291  and i really don t think you understand i came...      0   

                                                lemm_text  
0       explanation why the edit make under my usernam...  
1       d aww he match this background colour I m seem...  
2       hey man I m really not try to edit war it s ju...  
3       more I can t make any real suggestion on impro...  
4       you sir be my hero any chance you remember wha...  
...                                                   ...  
159287  and for the second time of ask when your view ...  
159288  you should be ashamed of yourself that be a ho...  
159289  spitzer umm there s no actual article for pros...  
159290  and it look like it be actually you who put on...  
159291  and I really don t think you understand I come...  

[159292 rows x 3 columns]

 <div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>
    
Очистка и лемматизация сделаны корректно)

## Обучение

In [7]:
X_train, X_test, y_train, y_test = train_test_split(df['lemm_text'], df['toxic'], test_size=.25, random_state=RANDOM_STATE)

In [8]:
vectorizer_tf = TfidfVectorizer(stop_words= set(nltk_stopwords.words('english')))
X_train_tf = vectorizer_tf.fit_transform(X_train)
X_test_tf = vectorizer_tf.transform(X_test)    

 <div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>
    
Разбиение было сделано верно. Отлично, что векторизатор был обучен только на тренировочной части данных
</div>

### Logistic Regression

Далее ячейки закоментил, т.к. расчет занимает много времени.

In [9]:
# log_reg_model = LogisticRegression(random_state=RANDOM_STATE)

# parameters = [
#     {'C' : [10, 20, 40, 80],
#     'max_iter' : [100, 1000, 2500, 5000]
#     } ]

# log_reg_grid = RandomizedSearchCV(log_reg_model, parameters, scoring='f1').fit(X_train_tf, y_train)
# log_reg_f1 = print(log_reg_grid.best_score_)
# print("Лучшая модель")
# print(log_reg_grid.best_estimator_)

In [10]:
model_lr = LogisticRegression(C=20, max_iter=2500, random_state=RANDOM_STATE)
model_lr.fit(X_train_tf, y_train)

lr_f1 = f1_score(y_train, model_lr.predict(X_train_tf))

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b>

        
Кросс-валидация одновремено с извлечением признаков из текстов требует особого подхода. Ты помнишь, что векторайзер мы обучали на обучающей выборке а применяли к тестовой . Что происходит при кросс-валидации? Ты разбиваешь большую обучающую выборку на несколько фолдов, из них выделяешь часть на обучение, часть на валидацию.  Но признаки (TF-IDF) уже обучены на всем наборе (обучающий + валидационный). Получается утечка данных. Чтобы избежать её нужно использовать pipeline из векторайзера и модели.
    
Пример такого пайплайна можно посмотреть в официальной документации
    
  https://scikit-learn.org/stable/auto_examples/model_selection/plot_grid_search_text_feature_extraction.html
    
И еще полезные материалы:
    
  https://newtechaudit.ru/pipeline-v-rabote-s-dannymi/           
        
        
        
</div>

### Random Forest

In [11]:
# model_rf = RandomForestClassifier(random_state=RANDOM_STATE)

# parameters = {'n_estimators': [8,16,32,64,128],
#               'max_depth': range(2,22, 4)}

# rf_grid = RandomizedSearchCV(model_rf, parameters, scoring='f1').fit(X_train_tf, y_train)
# rf_rmse = print(rf_grid.best_score_)

# print("Лучшая модель")
# print(rf_grid.best_estimator_)

In [12]:
model_rf = RandomForestClassifier(max_depth=14, n_estimators=32, random_state=RANDOM_STATE)
model_rf.fit(X_train_tf, y_train)

rf_f1 = f1_score(y_train, model_rf.predict(X_train_tf))

### Decision Tree

In [13]:
# model_dt = DecisionTreeClassifier(random_state=RANDOM_STATE)

# parameters = {'max_depth': range(10, 110, 10)}

# dt_grid = RandomizedSearchCV(model_dt, parameters, scoring='f1').fit(X_train_tf, y_train)
# dt_rmse = print(dt_grid.best_score_)

# print("Лучшая модель")
# print(dt_grid.best_estimator_)

In [14]:
model_dt = DecisionTreeClassifier(max_depth = 100, random_state=RANDOM_STATE)
model_dt.fit(X_train_tf, y_train)

dt_f1 = f1_score(y_train, model_dt.predict(X_train_tf))

### LightGBM

In [15]:
# model_lgbm = LGBMClassifier(random_state=RANDOM_STATE)

# parameters = {'max_depth': range (2, 10, 2),
#               'learning_rate': [0.1, 0.5, 0.8]}

# lgbm_grid = RandomizedSearchCV(model_lgbm, parameters, scoring='f1').fit(X_train_tf, y_train)
# lgbm_f1 = print(lgbm_grid.best_score_)
# print("Лучшая модель")
# print(lgbm_grid.best_estimator_)

In [16]:
model_lgbm = LGBMClassifier(learning_rate=0.5, max_depth=8, random_state=RANDOM_STATE)
model_lgbm.fit(X_train_tf, y_train)

lgbm_f1 = f1_score(y_train, model_lgbm.predict(X_train_tf))

 <div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b>
    
Молодец, что пробуешь разные модели в этом шаге)
</div>

## Анализ моделей

In [17]:
result = {}
for i, k in zip(['Logistic Regression', 'Random Forest', 'Decision Tree' ,'LightGBM'], ['lr', 'rf', 'dt','lgbm']):
    result[i] = [vars()[k + '_f1']]

result_table = pd.DataFrame(result.values(), index = result.keys(), columns = ['F1']  )
result_table.style.highlight_max(subset = ['F1'], color = 'lightgreen', axis = 0)

## Тестирование лучшей модели

In [18]:
f1_final = f1_score(y_test, model_lr.predict(X_test_tf))
print(f'F1 лучшей модели на тестовой выборке: {f1_final:.2f}')

F1 лучшей модели на тестовой выборке: 0.79


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
На тестовой выборке получено хорошее качество)

## Выводы

1. Была произведена загрузка данных и их подготовка.
2. Сформированы обучающая и тестовая выборки.
3. Были обучены 4 модели, на которых были подобраны оптимальные гиперпараметры. Наилучшая метрика F1 у модели LogisticRegression: 0.94.
5. Данная модель была протестирована на тестовой выборке и дала результат 0.79, что выше требуемого значения.

<div style="border:solid Chocolate 2px; padding: 40px">


<h2> Итоговый комментарий ревьюера <a class="tocSkip"> </h2>    
    
  
У меня сложилось хорошее общее впечатление о проекте, тебе удалось неплохо справиться с этим проектом. Молодец! Подготовлены данные для анализа, изучен каждый параметр. Все красиво визуализировано, построено несколько моделей классификации и оценено их качество. Осмысленная аналитика и дельная модельная работа - многое удалось как надо)
    
Отмечу отдельные положительные моменты проекта🙂:
    
- в ходе проекта встречались функции, помогающие избавиться от дублирования кода;
- хорошая интерпретация;
- при обучении моделей использована валидация и поиск гиперпараметров.
    
Проект может быть зачтен, но я его отправлю назад, чтобы у тебя была возможность задать вопросы и внести правки, при желании. Однако, ты можешь просто вернуть проект в таком же виде и я его зачту.
     
</div>

<div style="border:solid Chocolate 2px; padding: 40px">


<h2> Итоговый комментарий ревьюера 2 <a class="tocSkip"> </h2>    
    
     
  
Теперь почти идеально. Принимаю работу)
    
Если хочешь лучше разобраться в теме, то могу посоветовать тебе: 
    
https://huggingface.co/transformers/model_doc/bert.html
    
https://t.me/renat_alimbekov
    
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM
    
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены
    
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд по трансформеру от создателей pytorch
    
https://transformer.huggingface.co/ - поболтать с трансформером
    
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множествореализованных методов для трансформеров методов NLP
    
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html
    
    
    
В этом проекте были отработаны все моменты. 
    
Успехов тебе в следующем спринте!
    
Рад был помочь тебе)
    
</div>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны